compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp

```diff
@@ -7,20 +7,56 @@
 #include "iree/compiler/Codegen/Common/EncodingUtils.h"
 #include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenTypes.h"
 #include "iree/compiler/Codegen/Dialect/Codegen/Utils/Utils.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.h"
 #include "iree/compiler/Dialect/Encoding/IR/EncodingTypes.h"
 #include "iree/compiler/Dialect/TensorExt/IR/TensorExtTypes.h"
+#include "llvm/Support/Debug.h"
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
 #include "mlir/Dialect/Utils/IndexingUtils.h"
 #include "mlir/IR/BuiltinAttributes.h"
 
 #include <optional>
 
+#define DEBUG_TYPE "iree-codegen-encoding-utils"
+
 namespace mlir::iree_compiler {
 
 using IREE::Codegen::MaterializeEncodingInfo;
 using IREE::Encoding::EncodingAttr;
 using IREE::Encoding::getEncodingAttr;
 using IREE::Encoding::getEncodingContractionDims;
+using IREE::Encoding::PadEncodingLayoutAttr;
+using IREE::Encoding::SerializableEncodingAttrInterface;
+
+// Returns the layout as a SerializableEncodingAttrInterface, or nullptr if this
+// is not the only layout or if there's no encoding at all.
+static SerializableEncodingAttrInterface
+getSerializableEncodingAttr(Attribute layoutAttr, RankedTensorType type) {
+  if (!type.getEncoding()) {
+    return nullptr;
+  }
+  auto encoding =
+      dyn_cast_or_null<IREE::Encoding::LayoutAttr>(type.getEncoding());
+  if (encoding) {
+    ArrayAttr layouts = encoding.getLayouts();
+    if (layouts.size() != 1) {
+      return nullptr;
+    }
+    return dyn_cast<SerializableEncodingAttrInterface>(*layouts.begin());
+  }
+  auto encodingResolver =
+      dyn_cast<IREE::Encoding::EncodingLayoutResolverAttrInterface>(layoutAttr);
+  if (!encodingResolver) {
+    return nullptr;
+  }
+  Attribute resolvedEncoding = encodingResolver.getLayout(type);
+  LLVM_DEBUG({
+    llvm::dbgs() << "Unresolved type: " << type << "\n";
+    llvm::dbgs() << "layoutAttr: " << layoutAttr << "\n";
+    llvm::dbgs() << "Resolved into: " << resolvedEncoding << "\n";
+  });
```

Comment:
optional nit: we can define LDBG macro for this type of logs. E.g.,

https://github.com/iree-org/iree/blob/946b1eacf54e591e93c6567b000f4b57c02e7bd1/compiler/src/iree/compiler/Codegen/Common/MaterializeEncoding.cpp#L31-L33

Reply:
Yeah, thanks for the pointer. Updated it but could you check it's exactly how you had in mind?